File: backend/bridge.mjs
import { IMessageSDK } from "@photon-ai/imessage-kit";
import Database from "better-sqlite3";
import os from "os";
import path from "path";
import fs from "fs";

const BACKEND_URL = process.env.BACKEND_URL || "http://localhost:8000";
const POLL_MS = 500;
const DB_PATH = path.join(os.homedir(), "Library/Messages/chat.db");

const sdk = new IMessageSDK();
const db = new Database(DB_PATH, { readonly: true });

const attachmentStmt = db.prepare(`
  SELECT a.filename, a.mime_type
  FROM attachment a
  JOIN message_attachment_join maj ON a.ROWID = maj.attachment_id
  WHERE maj.message_id = ?
    AND a.mime_type LIKE 'image/%'
  LIMIT 1
`);

// Start from the current max ROWID so we only process new messages
let lastRowId = db.prepare("SELECT MAX(ROWID) as m FROM message").get().m ?? 0;
console.log(`🌉 iMessage bridge starting (last ROWID: ${lastRowId})...`);

async function poll() {
  const rows = db.prepare(`
    SELECT message.ROWID, message.text, handle.id as sender
    FROM message
    LEFT JOIN handle ON message.handle_id = handle.ROWID
    WHERE message.ROWID > ?
      AND message.is_from_me = 0
      AND message.item_type = 0
      AND (message.text IS NOT NULL OR message.cache_has_attachments = 1)
    ORDER BY message.ROWID ASC
  `).all(lastRowId);

  for (const row of rows) {
    lastRowId = row.ROWID;
    const text = row.text?.trim() || "";
    const sender = row.sender;
    if (!sender) continue;

    let image_base64 = null;
    let image_media_type = null;
    const attachment = attachmentStmt.get(row.ROWID);
    if (attachment) {
      const filePath = attachment.filename.replace("~", os.homedir());
      try {
        image_base64 = fs.readFileSync(filePath).toString("base64");
        image_media_type = attachment.mime_type;
        console.log(`🖼️  Image attachment from ${sender}: ${filePath}`);
      } catch (e) {
        console.error(`❌ Could not read attachment: ${e.message}`);
      }
    }

    if (!text && !image_base64) continue;

    console.log(`📩 From ${sender}: ${text || "(image only)"}`);

    try {
      const res = await fetch(`${BACKEND_URL}/api/bot/message`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ phone: sender, text, image_base64, image_media_type }),
      });
      const data = await res.json();
      if (data.reply) {
        await sdk.send(sender, data.reply);
        console.log(`📤 Replied: ${data.reply.slice(0, 80)}`);
      }
    } catch (err) {
      console.error("❌ Error:", err.message);
    }
  }
}

let busy = false;

async function safePoll() {
  if (busy) return;
  busy = true;
  try { await poll(); } finally { busy = false; }
}

console.log(`✅ Polling for new iMessages every ${POLL_MS / 1000}s...`);
setInterval(safePoll, POLL_MS);

File: backend/bridge.py
import os
import queue
import sqlite3
import tempfile
import threading
import time
import requests
import subprocess
from pathlib import Path

DB_PATH = Path.home() / "Library/Messages/chat.db"
BACKEND_URL = "http://localhost:8000"
POLL_INTERVAL = 0.5

_sender_queues: dict = {}
_sender_queues_lock = threading.Lock()


def get_connection():
    conn = sqlite3.connect(f"file:{DB_PATH}?mode=ro", uri=True)
    conn.row_factory = sqlite3.Row
    return conn


def get_latest_rowid(conn):
    row = conn.execute(
        "SELECT MAX(ROWID) as max_id FROM message WHERE is_from_me = 0"
    ).fetchone()
    return row["max_id"] or 0


def get_new_messages(conn, since_rowid):
    return conn.execute("""
        SELECT m.ROWID, m.text, m.is_from_me, m.date,
               h.id as sender, c.chat_identifier
        FROM message m
        LEFT JOIN handle h ON m.handle_id = h.ROWID
        LEFT JOIN chat_message_join cmj ON m.ROWID = cmj.message_id
        LEFT JOIN chat c ON cmj.chat_id = c.ROWID
        WHERE m.ROWID > ?
          AND m.is_from_me = 0
          AND m.text IS NOT NULL
          AND m.text != ''
        ORDER BY m.ROWID ASC
    """, (since_rowid,)).fetchall()


def send_reply(sender, reply_text):
    with tempfile.NamedTemporaryFile(mode="w", suffix=".txt", delete=False) as f:
        f.write(reply_text)
        tmp_path = f.name
    try:
        script = f"""
set msgText to (do shell script "cat " & quoted form of "{tmp_path}")
tell application "Messages"
    set myService to first service whose service type = iMessage
    set myBuddy to buddy "{sender}" of myService
    send msgText to myBuddy
end tell
"""
        subprocess.run(["osascript", "-e", script], check=True)
    finally:
        os.unlink(tmp_path)


def sender_worker(sender, q):
    while True:
        item = q.get()
        if item is None:
            break
        text, chat_guid = item
        try:
            res = requests.post(f"{BACKEND_URL}/api/bot/message", json={
                "phone": sender,
                "chat_guid": chat_guid or sender,
                "text": text,
            }, timeout=90)
            data = res.json()
            if data.get("reply"):
                send_reply(sender, data["reply"])
                print(f"Replied to {sender}: {data['reply'][:80]}")
        except Exception as e:
            print(f"Error handling message from {sender}: {e}")


def enqueue_message(sender, text, chat_guid):
    with _sender_queues_lock:
        if sender not in _sender_queues:
            q = queue.Queue()
            _sender_queues[sender] = q
            t = threading.Thread(target=sender_worker, args=(sender, q), daemon=True)
            t.start()
        _sender_queues[sender].put((text, chat_guid))


def main():
    print("iMessage bridge starting...")
    conn = get_connection()
    last_rowid = get_latest_rowid(conn)
    conn.close()
    print(f"Watching for new messages (last ROWID: {last_rowid})...")

    while True:
        try:
            conn = get_connection()
            new_msgs = get_new_messages(conn, last_rowid)
            conn.close()

            for msg in new_msgs:
                last_rowid = max(last_rowid, msg["ROWID"])
                sender = msg["sender"]
                text = msg["text"]
                if not sender or not text:
                    continue

                print(f"From {sender}: {text}")
                enqueue_message(sender, text, msg["chat_identifier"])

        except Exception as e:
            print(f"DB error: {e}")

        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    main()
